nova/virt/vmwareapi/vm_util.py

```diff
@@ -1513,79 +1515,176 @@ def _get_host_reservations_map(groups=None):
+        total_memory_mb += memory_mb
+        memory_mb_used = stats["memory_mb_used"]
+        total_memory_mb_used += memory_mb_used
+        memory_mb_reserved = stats["memory_mb_reserved"]
+        total_memory_mb_reserved += memory_mb_reserved
+        max_mem_mb_per_host = max(max_mem_mb_per_host,
+                                  memory_mb - memory_mb_reserved)
+
+        if cpu_infos_equal and previous_cpu_info:
+            cpu_infos_equal = (previous_cpu_info == stats["cpu_info"])
+
+        previous_cpu_info = stats["cpu_info"]
+
+    # Calculate VM-reservable memory as a ratio of total available
+    # memory, depending on either the configured tolerance for failed
+    # hypervisors or a single configurable ratio.
+    max_fail_hvs = \
+        CONF.vmware.memory_reservation_cluster_hosts_max_fail
+    if max_fail_hvs and total_hypervisor_count:
+        vm_reservable_memory_ratio = \
+            (1 - max_fail_hvs / total_hypervisor_count)
+    else:
+        vm_reservable_memory_ratio = \
+            CONF.vmware.memory_reservation_max_ratio_fallback
+
+    return {
+        "vcpus": total_vcpus,
+        "vcpus_used": total_vcpus_used,
+        "vcpus_reserved": total_vcpus_reserved,
+        "max_vcpus_per_host": max_vcpus_per_host,
+        "memory_mb": total_memory_mb,
+        "memory_mb_used": total_memory_mb_used,
+        "memory_mb_reserved": total_memory_mb_reserved,
+        "max_mem_mb_per_host": max_mem_mb_per_host,
+        "vm_reservable_memory_ratio": vm_reservable_memory_ratio,
+        "cpu_info": (previous_cpu_info if cpu_infos_equal
+                     else "CPUs for this cluster have different values!")
+    }
+
+
+def _host_props_to_cpu_info(host_props):
+    processor_type = None
+    cpu_vendor = None
+    hardware_cpu_pkg = host_props.get("hardware.cpuPkg")
+    if hardware_cpu_pkg and hardware_cpu_pkg.HostCpuPackage:
+        t = hardware_cpu_pkg.HostCpuPackage[0]
+        processor_type = t.description
+        cpu_vendor = t.vendor.title()
+
+    features = []
+    if "config.featureCapability" in host_props:
+        feature_capability = host_props["config.featureCapability"]
+        for feature in feature_capability.HostFeatureCapability:
+            if not feature.featureName.startswith("cpuid."):
+                continue
+            if feature.value != "1":
+                continue
+
+            name = feature.featureName
+            features.append(name.split(".", 1)[1].lower())
+    cpu_info = {
+        "model": processor_type,
+        "vendor": cpu_vendor,
+        "features": sorted(features)
+    }
+    hardware_cpu_info = host_props.get("hardware.cpuInfo")
+    if hardware_cpu_info:
+        cpu_info["topology"] = {
+            "cores": hardware_cpu_info.numCpuCores,
+            "sockets": hardware_cpu_info.numCpuPackages,
+            "threads": hardware_cpu_info.numCpuThreads
+        }
+    return cpu_info
+
+
+def _parse_product(stats, host_props):
```

Comment:
Not sure if `_parse_product()` is the right name here. It doesn't seem to parse anything really and imho should be a `_set_something()` instead, because it changes the given `stats`.

Reply:
I'll rename it to `_set_hypervisor_type_and_version`